From 4a7b560f9b4e960a2a3046bcbf84716eea63035d Mon Sep 17 00:00:00 2001
From: Totto16 <tobiausgais@gmail.com>
Date: Tue, 25 Aug 2026 17:35:20 +0200
Subject: [PATCH 3/5] feat: add support for C and C++ constructors and
 destructors - this requires as linker script change and GenFW modifications -
 this can 5than be used by edk2-libc and libcxx for uefi

---
 BaseTools/Scripts/GccBase.lds           | 61 +++++++++++++++++++++++++
 BaseTools/Source/C/GenFw/Elf32Convert.c | 35 ++++++++++++++
 BaseTools/Source/C/GenFw/Elf64Convert.c | 35 ++++++++++++++
 3 files changed, 131 insertions(+)

diff --git a/BaseTools/Scripts/GccBase.lds b/BaseTools/Scripts/GccBase.lds
index 3d2a97ddeb..32b54385fc 100644
--- a/BaseTools/Scripts/GccBase.lds
+++ b/BaseTools/Scripts/GccBase.lds
@@ -85,6 +85,67 @@ SECTIONS {
    */
   .build-id (INFO) : { *(.note.gnu.build-id) }
 
+  /*
+   * Retain the GCC static / global constructor / initializer and destructors / finalizer
+   */
+  .init_array : {
+    PROVIDE_HIDDEN(__init_array_start = .);
+
+    KEEP(*(SORT_BY_INIT_PRIORITY(.init_array.*)))
+    KEEP(*(.init_array))
+
+    PROVIDE_HIDDEN(__init_array_end = .);
+
+    PROVIDE_HIDDEN(__libc_init_array_start = .);
+
+    KEEP(*(SORT_BY_NAME(.__libc_init.*)))
+
+    PROVIDE_HIDDEN(__libc_init_array_end = .);
+  }
+
+  .fini_array : {
+    PROVIDE_HIDDEN(__fini_array_start = .);
+
+    KEEP(*(SORT_BY_INIT_PRIORITY(.fini_array.*)))
+    KEEP(*(.fini_array))
+
+    PROVIDE_HIDDEN(__fini_array_end = .);
+
+    PROVIDE_HIDDEN(__libc_fini_array_start = .);
+
+    KEEP(*(SORT_BY_NAME(.__libc_fini.*)))
+
+    PROVIDE_HIDDEN(__libc_fini_array_end = .);
+  }
+
+  /*
+   * Error on unsupported (or not yet implemented) sections for construction / destruction
+   */
+  .preinit_array : {
+    *(.preinit_array)
+  }
+  ASSERT(SIZEOF(.preinit_array) == 0, "error: .preinit_array is not supported (yet)")
+
+  .init : {
+    *(.init)
+  }
+  ASSERT(SIZEOF(.init) == 0, "error: .init is deprecated (System V uses it, and GCC uses other mechanisms)")
+
+  .fini : {
+    *(.fini)
+  }
+  ASSERT(SIZEOF(.fini) == 0, "error: .fini is deprecated (System V uses it, and GCC uses other mechanisms)")
+
+  .ctors : {
+    *(.ctors)
+  }
+  ASSERT(SIZEOF(.ctors) == 0, "error: .ctors is not implemented yet (C++ uses that)")
+
+  .dtors : {
+    *(.dtors)
+  }
+  ASSERT(SIZEOF(.dtors) == 0, "error: .dtors is not implemented yet (C++ uses that)")
+
   /DISCARD/ : {
     *(.note.GNU-stack)
     *(.gnu_debuglink)
diff --git a/BaseTools/Source/C/GenFw/Elf32Convert.c b/BaseTools/Source/C/GenFw/Elf32Convert.c
index 7b4fb4410a..45d5965228 100644
--- a/BaseTools/Source/C/GenFw/Elf32Convert.c
+++ b/BaseTools/Source/C/GenFw/Elf32Convert.c
@@ -239,6 +239,26 @@ IsHiiRsrcShdr (
   return (BOOLEAN) (strcmp((CHAR8*)mEhdr + Namedr->sh_offset + Shdr->sh_name, ELF_HII_SECTION_NAME) == 0);
 }
 
+#define ELF_INIT_ARRAY_SECTION_NAME  ".init_array"
+#define ELF_FINI_ARRAY_SECTION_NAME  ".fini_array"
+
+STATIC
+BOOLEAN
+IsInitFiniArrayShdr (
+  Elf_Shdr *Shdr
+  )
+{
+  Elf_Shdr *Namedr = GetShdrByIndex(mEhdr->e_shstrndx);
+  const CHAR8 *Name = (const CHAR8 *)mEhdr +
+         Namedr->sh_offset +
+         Shdr->sh_name;
+
+  return (BOOLEAN)(
+    strcmp(Name, ELF_INIT_ARRAY_SECTION_NAME) == 0 ||
+    strcmp(Name, ELF_FINI_ARRAY_SECTION_NAME) == 0
+  );
+}
+
 STATIC
 BOOLEAN
 IsDataShdr (
@@ -248,6 +268,9 @@ IsDataShdr (
   if (IsHiiRsrcShdr(Shdr)) {
     return FALSE;
   }
+  if (IsInitFiniArrayShdr (Shdr)) {
+    return TRUE;
+  }
   return (BOOLEAN) (Shdr->sh_flags & (SHF_EXECINSTR | SHF_WRITE | SHF_ALLOC)) == (SHF_ALLOC | SHF_WRITE);
 }
 
@@ -662,6 +685,18 @@ WriteSections32 (
         memset(mCoffFile + mCoffSectionsOffset[Idx], 0, Shdr->sh_size);
         break;
 
+      case SHT_INIT_ARRAY:
+      case SHT_FINI_ARRAY:
+        //NOTE: do the same as the SHT_PROGBITS section (we SHOULd have already determined them as data sections, so there should be space)
+        /* Copy.  */
+        if (Shdr->sh_offset + Shdr->sh_size > mFileBufferSize) {
+          return FALSE;
+        }
+        memcpy(mCoffFile + mCoffSectionsOffset[Idx],
+              (UINT8*)mEhdr + Shdr->sh_offset,
+              (size_t) Shdr->sh_size);
+        break;
+
       default:
         //
         //  Ignore for unknown section type.
diff --git a/BaseTools/Source/C/GenFw/Elf64Convert.c b/BaseTools/Source/C/GenFw/Elf64Convert.c
index 1208137bca..c1a01de706 100644
--- a/BaseTools/Source/C/GenFw/Elf64Convert.c
+++ b/BaseTools/Source/C/GenFw/Elf64Convert.c
@@ -323,6 +323,26 @@ IsSymbolShdr (
   return (BOOLEAN) (strcmp((CHAR8*)mEhdr + Namehdr->sh_offset + Shdr->sh_name, ELF_SYMBOL_SECTION_NAME) == 0);
 }
 
+#define ELF_INIT_ARRAY_SECTION_NAME  ".init_array"
+#define ELF_FINI_ARRAY_SECTION_NAME  ".fini_array"
+
+STATIC
+BOOLEAN
+IsInitFiniArrayShdr (
+  Elf_Shdr *Shdr
+  )
+{
+  Elf_Shdr *Namedr = GetShdrByIndex(mEhdr->e_shstrndx);
+  const CHAR8 *Name = (const CHAR8 *)mEhdr +
+         Namedr->sh_offset +
+         Shdr->sh_name;
+
+  return (BOOLEAN)(
+    strcmp(Name, ELF_INIT_ARRAY_SECTION_NAME) == 0 ||
+    strcmp(Name, ELF_FINI_ARRAY_SECTION_NAME) == 0
+  );
+}
+
 STATIC
 BOOLEAN
 IsDataShdr (
@@ -332,6 +352,9 @@ IsDataShdr (
   if (IsHiiRsrcShdr(Shdr)) {
     return FALSE;
   }
+  if (IsInitFiniArrayShdr (Shdr)) {
+    return TRUE;
+  }
   return (BOOLEAN) (Shdr->sh_flags & (SHF_EXECINSTR | SHF_WRITE | SHF_ALLOC)) == (SHF_ALLOC | SHF_WRITE);
 }
 
@@ -1388,6 +1411,18 @@ WriteSections64 (
         memset(mCoffFile + mCoffSectionsOffset[Idx], 0, (size_t) Shdr->sh_size);
         break;
 
+      case SHT_INIT_ARRAY:
+      case SHT_FINI_ARRAY:
+        //NOTE: do the same as the SHT_PROGBITS section (we SHOULd have already determined them as data sections, so there should be space)
+        /* Copy.  */
+        if (Shdr->sh_offset + Shdr->sh_size > mFileBufferSize) {
+          return FALSE;
+        }
+        memcpy(mCoffFile + mCoffSectionsOffset[Idx],
+              (UINT8*)mEhdr + Shdr->sh_offset,
+              (size_t) Shdr->sh_size);
+        break;
+
       default:
         //
         //  Ignore for unknown section type.
-- 
2.55.0

